### PR TITLE
`/blocks/:id` page

### LINKED ISSUE #158: Implement `/blocks/:id` Page
#### What would you like to be added:

`[id]` can be height or hash, on page router, it must be replaced with hash.

```txt
/src/pages/blocks/[id].tsx
```

/area blocks transactions
/triage accepted
/assign @aikchun

### PR DESCRIPTION
**Pull request by aikchun**
#### What kind of PR is this?:

/kind feature

#### Which issue(s) does this PR fixes?:

Fixes #158 

#### Additional comments?:
Missing features:
- [x] transaction timestamp
- [x] previous block
- [x] next block
- [ ] block reward create ticket on upstream https://github.com/DeFiCh/whale/issues/383
- [ ] ~~search for the transactions table~~ not doing
- [ ] ~~filter for the transactions table~~ not doing
---

- Added `TZ=UTC` to the `npm run cypress:run` command so that cypress runs on UTC just like the CI.

Screenshot:
image

### CODE REVIEW

File: src/utils/index.ts [hunk 1]
```diff
@@ -0,0 +1 @@
+export * from './parseAge'
```
Comment (fuxingloh) [inline src/utils/index.ts:1]: replace with https://github.com/DeFiCh/scan/blob/b93efa3cec97a469ad5ce499fa6b8ae75ff822ee/src/pages/blocks/index.tsx#L8

File: src/components/Table/index.tsx [hunk 2]
```diff
@@ -0,0 +1,2 @@
+export * from './TableHeader'
+export * from './TableBody'
```
Comment (fuxingloh) [inline src/components/Table/index.tsx:2]: don't use index, hard to maintain and this is not library project

File: src/components/Table/TableHeader.tsx [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+export interface TableProps {
+  className?: string
+  children: ReactNode
+}
+
+export function TableHeader (props: TableProps): JSX.Element {
+  const { className = '', children } = props
+  return (
+    <div className={`table-header flex ${className}`}>
```
Comment (fuxingloh) [inline src/components/Table/TableHeader.tsx:11]: use `classNames`, see other component that uses it

File: src/components/Table/TableBody.tsx [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+  return (
+    <div className={`table-body flex flex-col ${className}`}>
+      {children}
+    </div>
+  )
+}
+
+TableBody.defaulProps = {
+  className: ''
+}
```
Comment (fuxingloh) [inline src/components/Table/TableBody.tsx:19]: remove this and use classNames 

https://github.com/DeFiCh/scan/blob/b93efa3cec97a469ad5ce499fa6b8ae75ff822ee/src/components/commons/AdaptiveTable.tsx#L23

File: src/components/Table/TableBody.tsx [hunk 5]
```diff
@@ -0,0 +1,19 @@
+import { ReactNode } from 'react'
+
+interface TableBodyProps {
+  className?: string
+  children: ReactNode
+}
+
+export function TableBody (props: TableBodyProps): JSX.Element {
```
Comment (fuxingloh) [inline src/components/Table/TableBody.tsx:8]: For simple classes like this, too simple without any functionality, don't create a component for it.

File: src/pages/blocks/[blockId]/transactions/index.tsx [hunk 6]
```diff
@@ -0,0 +1,54 @@
+import { useRouter } from 'next/router'
+import { TableHeader, TableBody } from '../../../../components/Table/index'
+
+interface Transaction {
```
Comment (fuxingloh) [inline src/pages/blocks/[blockId]/transactions/index.tsx:4]: Please import these from whale, and don't re-create your own.

File: src/pages/blocks/index.tsx [hunk 7]
```diff
@@ -20,7 +20,6 @@ export default function Blocks ({ blocks }: InferGetServerSidePropsType<typeof g
   return (
     <div className='container mx-auto px-4 pt-12 pb-20'>
       <Head title='Blocks' />
-
```
Comment (fuxingloh) [inline src/pages/blocks/index.tsx:23]: remove this change to prevent upstream conflicts

File: tailwind.config.js [hunk 8]
```diff
@@ -7,18 +7,20 @@ module.exports = {
...
       colors: {
-        primary: '#ff00af',
+        primary: '#ff00af'
       }
     }
   },
   variants: {
-    extend: {}
+    extend: {
+      borderRadius: ['first', 'last']
```
Comment (fuxingloh) [inline tailwind.config.js:20]: what is this for, if not using, revert it

**Review by fuxingloh — CHANGES_REQUESTED** [review]

File: src/pages/blocks/[blockId]/transactions/index.tsx (hunk 6 above)
Comment (aikchun) [reply src/pages/blocks/[blockId]/transactions/index.tsx:4]: Sorry this no applicable. The changes for this was deleted in the blocks page. But it wasn't done for this branch.

File: src/components/Table/TableBody.tsx (hunk 5 above)
Comment (aikchun) [reply src/components/Table/TableBody.tsx:8]: This is not applicable anymore.

File: tailwind.config.js (hunk 8 above)
Comment (aikchun) [reply tailwind.config.js:20]: This is used in my `renderBlockDetails`
```
      <div className='px-6 py-4 border first:rounded-t-md last:rounded-b-md flex justify-between' key={d.label}>
        <span className='w-1/3 flex-shrink-0'>
          {d.label}
        </span>
```
The plugin needs to be enabled according to tailwindcss [docs](https://tailwindcss.com/docs/hover-focus-and-other-states#first-child)

File: src/utils/index.ts (hunk 1 above)
Comment (aikchun) [reply src/utils/index.ts:1]: This utility function is deleted.

File: src/pages/blocks/index.tsx (hunk 7 above)
Comment (aikchun) [reply src/pages/blocks/index.tsx:23]: Done.

File: src/components/Table/TableBody.tsx (hunk 4 above)
Comment (aikchun) [reply src/components/Table/TableBody.tsx:19]: Deleted this component.

File: src/components/Table/TableHeader.tsx (hunk 3 above)
Comment (aikchun) [reply src/components/Table/TableHeader.tsx:11]: Deleted this component.

File: tailwind.config.js (hunk 8 above)
Comment (fuxingloh) [reply tailwind.config.js:20]: don't really need this, border-radius parent box, and divider between the sub sections

File: src/pages/blocks/[blockId]/index.tsx [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+      label: 'Version:', value: `${block.version}`, testId: 'block-detail-version'
+    }
+    // {
+    //   label: 'Next Block:', value: `${block.nextBlock.height}`
+    // },
+    // {
+    //   label: 'Previous Block:', value: `${block.prevBlock.height}`
+    // }
+
+  ]
```
Comment (fuxingloh) [inline src/pages/blocks/[blockId]/index.tsx:112]: Don't do this, Create a Component instead

for each Component, don't for each the HTML

File: src/pages/blocks/[blockId]/index.tsx [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+
+  function renderBlockDetails (details: Array<{ label: string, content?: JSX.Element, value?: string, testId?: string }>): JSX.Element[] {
```
Comment (fuxingloh) [inline src/pages/blocks/[blockId]/index.tsx:114]: Don't pass Array into a component and create a list of Arrayed Element[]

Loop the Array instead and map the Component.

File: tailwind.config.js (hunk 8 above)
Comment (fuxingloh) [reply tailwind.config.js:20]: Revert this file, best not to extend features that we don't need.

**Review by fuxingloh — CHANGES_REQUESTED** [review]

File: src/pages/blocks/[blockId]/index.tsx (hunk 9 above)
Comment (fuxingloh) [reply src/pages/blocks/[blockId]/index.tsx:112]: I realized my sentences can be misleading, in terms of `Component`, avoid creating shared global components when you don't have to. However, always opt to better organize your code via internal Component within the same file.

For `array.map()` avoid passing an array and returning `JSX.Element[]`. However, you should `function Component(props: {}): JSX.Element`. Functional component in React has special property, you should always opt to use it. For more information, take a look at this issue: https://github.com/DeFiCh/wallet/issues/694

File: src/pages/blocks/[blockId]/index.tsx [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+        }
+      </div>
+    ))
+  }
+
+  return (
+    <div className='container mx-auto px-4 py-8'>
+      <div className='flex items-center justify-center pb-6'>
+        <div className='bg-pink-50 rounded p-3'>
+          🚧 Work in progress, this is an early iteration of defiscan.live/blocks/[blockHash]. Some features are not available and
```
Comment (monstrobishi) [inline src/pages/blocks/[blockId]/index.tsx:137]:
```suggestion
          🚧 Work in progress, this is an early iteration of defiscan.live/blocks/[blockHeight|blockHash]. Some features are not available and
```

File: src/pages/blocks/[blockId]/index.tsx [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+import { OverflowTable } from '@components/commons/OverflowTable'
+import { CursorPage, CursorPagination } from '@components/commons/CursorPagination'
+import { getWhaleApiClient } from '@contexts/WhaleContext'
+import { IoCopyOutline } from 'react-icons/io5'
+
+import { fromUnixTime, format } from 'date-fns'
+
+interface BlockDetailsPageProps {
+  block: Block
+  confirmations: number
```
Comment (fuxingloh) [inline src/pages/blocks/[blockId]/index.tsx:17]:
```suggestion
```

File: src/pages/blocks/[blockId]/index.tsx [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+    },
+    {
+      label: 'Merkle Root:',
+      children:
+  <>
+    <span className='flex-grow break-all' data-testid='block-detail-merkle-root'>{block.merkleroot}</span>
+    <CopyButton text={block.merkleroot} />
+  </>
+    }
+  ]
```
Comment (fuxingloh) [inline src/pages/blocks/[blockId]/index.tsx:115]: Turn this into a natural HTML hierarchy, don't do any functional transformation on static data. It just makes code harder to maintain, and harder to trace.

```html
<BlockDetail/>
<BlockDetail/>
<BlockDetail/>
<BlockDetail/>
```

File: src/pages/blocks/[blockId]/index.tsx [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+      label: 'Version:', text: `${block.version}`, testId: 'block-detail-version'
+    },
+    {
+      label: 'Next Block:', text: 'next block (placeholder)'
+    },
+    {
+      label: 'Previous Block:', text: 'previous block (placeholder)'
+    }
+
+  ]
```
Comment (fuxingloh) [inline src/pages/blocks/[blockId]/index.tsx:137]: Same for this.

> Turn this into a natural HTML hierarchy, don't do any functional transformation on static data. It just makes code harder to maintain, and harder to trace.

File: src/pages/blocks/[blockId]/index.tsx [hunk 15]
```diff
@@ -0,0 +1,221 @@
...
+      <div className='flex items-center justify-center pb-6'>
+        <div className='bg-pink-50 rounded p-3'>
+          🚧 Work in progress, this is an early iteration of defiscan.live/blocks/[blockHash]. Some features are not available and
+          may not work as expected.
+        </div>
+      </div>
+      <Breadcrumb items={[
+        { path: '/blocks', name: 'Blocks' },
+        { path: `/blocks/${block.height}`, name: `#${block.height}`, canonical: true }
+
```
Comment (fuxingloh) [inline src/pages/blocks/[blockId]/index.tsx:150]:
```suggestion
```

File: src/pages/blocks/[blockId]/index.tsx [hunk 16]
```diff
@@ -0,0 +1,221 @@
...
+      ]}
+      />
+
+      <h1 className='font-semibold text-2xl'>Block #{block.height}</h1>
+      <div className='flex items-center'><span className='font-semibold'>Hash:&nbsp;</span> <span className='text-primary' data-testid='block-hash'>{block.hash}</span> <CopyButton text={block.hash} /></div>
+      <div className='flex mt-6 gap-x-6'>
+        <div className='flex flex-col w-5/12 flex-grow rounded border divide-solid divide-y-2'>
```
Comment (fuxingloh) [inline src/pages/blocks/[blockId]/index.tsx:157]:
```suggestion
        <div className='flex flex-col w-5/12 flex-grow rounded-md border divide-solid divide-y-1'>
```

File: src/pages/blocks/[blockId]/index.tsx [hunk 17]
```diff
@@ -0,0 +1,221 @@
...
+          </OverflowTable.Header>
+
+          {transactions.items.map(transaction => {
+            return (
+              <OverflowTable.Row key={transaction.hash}>
+                <OverflowTable.Cell>
+                  {transaction.hash}
+                </OverflowTable.Cell>
+                <OverflowTable.Cell>
+                  timestamp (placeholder)
```
Comment (fuxingloh) [inline src/pages/blocks/[blockId]/index.tsx:183]: This is same as block time

File: src/pages/blocks/[blockId]/index.tsx [hunk 18]
```diff
@@ -0,0 +1,221 @@
...
+                </OverflowTable.Cell>
+                <OverflowTable.Cell>
+                  {confirmations}
```
Comment (fuxingloh) [inline src/pages/blocks/[blockId]/index.tsx:186]:
```ts
const { count: { blocks } } = useSelector((state: RootState) => state.stats)
const confirmations = blocks - block.height
```

File: src/pages/blocks/[blockId]/index.tsx [hunk 19]
```diff
@@ -0,0 +1,221 @@
...
+            leftBlockDetails.map(detail => <BlockDetail {...detail} key={detail.label} />)
+          }
+        </div>
+        <div className='flex flex-col w-5/12 flex-grow rounded border divide-solid divide-y-2'>
+          {
+            rightBlockDetails.map(detail => <BlockDetail {...detail} key={detail.label} />)
+          }
+        </div>
+      </div>
+      <div className='my-6'>
```
Comment (fuxingloh) [inline src/pages/blocks/[blockId]/index.tsx:168]: image

Where is this?

File: src/pages/blocks/[blockId]/index.tsx [hunk 20]
```diff
@@ -0,0 +1,221 @@
...
+      </span>
+      {
+        (children !== undefined)
+          ? children
+          : (
+            <span className='flex-grow' data-testid={testId}>
+              {text}
+            </span>
+          )
+      }
```
Comment (fuxingloh) [inline src/pages/blocks/[blockId]/index.tsx:69]:
```suggestion
      <div className='w-1/3 flex-shrink-0'>
        {label}
      </div>
      <div className='flex-grow' data-testid={testId}>
        {children}
      </div>
```

File: src/pages/blocks/[blockId]/index.tsx [hunk 21]
```diff
@@ -0,0 +1,221 @@
...
+              Copied!
+            </span>
+          )
+          : null
+      }
+    </div>
+  )
+}
+
+function BlockDetail (props: { label: string, text: string, children?: ReactNode, testId?: string }): JSX.Element {
```
Comment (fuxingloh) [inline src/pages/blocks/[blockId]/index.tsx:54]:
```suggestion
function BlockDetail (props: PropsWithChildren<{...}>): JSX.Element {
```

**Review by fuxingloh — CHANGES_REQUESTED** [review]
Please match design 1 to 1

File: src/pages/blocks/[blockId]/index.tsx (hunk 21 above)
Comment (aikchun) [reply src/pages/blocks/[blockId]/index.tsx]: Made this change

File: src/pages/blocks/[blockId]/index.tsx (hunk 20 above)
Comment (aikchun) [reply src/pages/blocks/[blockId]/index.tsx]: Changed.

File: src/pages/blocks/[blockId]/index.tsx (hunk 19 above)
Comment (aikchun) [reply src/pages/blocks/[blockId]/index.tsx]: Added

File: src/pages/blocks/[blockId]/index.tsx (hunk 18 above)
Comment (aikchun) [reply src/pages/blocks/[blockId]/index.tsx:129]: Changed.

File: src/pages/blocks/[blockId]/index.tsx (hunk 17 above)
Comment (aikchun) [reply src/pages/blocks/[blockId]/index.tsx]: Changed.

File: src/pages/blocks/[blockId]/index.tsx (hunk 16 above)
Comment (aikchun) [reply src/pages/blocks/[blockId]/index.tsx]: Changed.  Used `divide-y` as there's no `divide-y-1`

File: src/pages/blocks/[blockId]/index.tsx (hunk 15 above)
Comment (aikchun) [reply src/pages/blocks/[blockId]/index.tsx]: Removed.

File: src/pages/blocks/[blockId]/index.tsx (hunk 14 above)
Comment (aikchun) [reply src/pages/blocks/[blockId]/index.tsx]: Changed

File: src/pages/blocks/[blockId]/index.tsx (hunk 13 above)
Comment (aikchun) [reply src/pages/blocks/[blockId]/index.tsx]: Changed.

File: src/pages/blocks/[blockId]/index.tsx (hunk 12 above)
Comment (aikchun) [reply src/pages/blocks/[blockId]/index.tsx]: Removed.

File: src/pages/blocks/[blockId]/index.tsx [hunk 22]
```diff
@@ -0,0 +1,221 @@
...
+          ? (
+            <span className='bg-gray-100 p-1 rounded absolute mr-4'>
+              Copied!
+            </span>
+          )
+          : null
+      }
+    </div>
+  )
+}
```
[the remark this answers is not in the document]
Comment (aikchun) [reply src/pages/blocks/[blockId]/index.tsx]: It's moved to components/common/CopyButton.tsx

File: src/pages/blocks/[blockId]/index.tsx [hunk 23]
```diff
@@ -0,0 +1,221 @@
...
+      }
+    </div>
+
+  )
+}
+
+BlockDetail.defaultProps = {
+  text: '',
+  label: ''
+}
```
[the remark this answers is not in the document]
Comment (aikchun) [reply src/pages/blocks/[blockId]/index.tsx]: Removed.

File: src/pages/blocks/[blockId]/index.tsx [hunk 24]
```diff
@@ -0,0 +1,221 @@
...
+  const block = await api.blocks.get(blockId)
+
+  const next = CursorPagination.getNext(context)
+  const transactions = await api.blocks.getTransactions(block.id, 50, next)
+  const blocks = await api.blocks.list()
+
+  return {
+    props: {
+      block,
+      confirmations: blocks[0].height - block.height,
```
[the remark this answers is not in the document]
Comment (aikchun) [reply src/pages/blocks/[blockId]/index.tsx]: Removed

File: src/pages/blocks/[blockId]/index.tsx [hunk 25]
```diff
@@ -0,0 +1,221 @@
...
+      </div>
+      <div className='flex justify-end mt-8'>
+        <CursorPagination pages={transactions.pages} path={`/blocks/${block.hash}`} />
+      </div>
+    </div>
+  )
+}
+
+// first of all I need to fetch the block
+//
```
[the remark this answers is not in the document]
Comment (aikchun) [reply src/pages/blocks/[blockId]/index.tsx]: Removed.

**Review by siradji — DISMISSED** [review]

**Comment by aikchun** [comment]
@fuxingloh I have rebased this branch and fixed the tests.

**Review by fuxingloh — APPROVED** [review]

**Pull request merged**